### PR TITLE
Draft the Schedule object

### PR DESCRIPTION
**Pull request by Magomogo**

### CODE REVIEW

File: src/Schedule.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * @param \DateTime $from
+     * @param \DateInterval $period
+     *
+     * @return \DateTime[]
+     */
+    public function materialize(\DateTime $from, \DateInterval $period): array
+    {
+        $to = \DateTime::createFromFormat('Y-m-d H:i', $from->format('Y-m-d H:i'));
```
Comment (Magomogo) [inline src/Schedule.php]: Use `$to =(clone $from)->add($period)`

File: src/Schedule.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        $to->add($period);
+        $schedule = [];
+        if ($from->format('l') != $this->rules[0]->getWeekday()) {
+            $from->modify('next ' . $this->rules[0]->getWeekday());
```
Comment (Magomogo) [inline src/Schedule.php]: The knowledge about weekdays should be encapsulated into `Schedule\Weekday` object. You need to correctly define `Schedule\RuleInterface` to make sure no implementation details will leak outside.

I propose public methods:

```

public function includedDays(\DateTime $from, \DateInterval $period): array; // DateTime[] here

public function excludedDays(\DateTime $from, \DateInterval $period): array; // DateTime[] here

```

File: src/Schedule/Weekday.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+declare(strict_types=1);
+
+namespace GroupLife\Core\Schedule;
+
+class Weekday implements RuleInterface
+{
+    private string $weekday;
+    private string $startTime;
+    private \DateInterval $period;
```
Comment (Magomogo) [inline src/Schedule/Weekday.php]: Not all actual PHP versions support typed class properties, please correct.

File: src/Schedule.php (hunk 1 above)
Comment (Lehersun) [reply src/Schedule.php]: Already fixed

File: src/Schedule.php (hunk 2 above)
Comment (Lehersun) [reply src/Schedule.php]: Pls check new commit

File: src/Schedule/RuleInterface.php [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace GroupLife\Core\Schedule;
+
+interface RuleInterface
+{
+    public function includedDays(\DateTime $from, \DateInterval $period, array $schedule): array;
+
+    public function excludedDays(\DateTime $from, \DateInterval $period, array $schedule): array;
```
Comment (Magomogo) [inline src/Schedule/RuleInterface.php]: Let's start here.

Not clear why this 'array $schedule' parameter is exists. Is it some list of dates? You don't need it to list excluded or included days according to some rule.

File: src/Schedule/WeekdayRule.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        if ($from->format('l') != $this->weekday) {
+            $from->modify('next ' . $this->weekday);
+        }
+        $from->modify($this->startTime);
+
+        while ($from <= $to) {
+            $schedule[] = clone $from;
+            $from->add($this->repeat);
+        }
+        return $schedule;
```
Comment (Magomogo) [inline src/Schedule/WeekdayRule.php:33]: Here you should just create array of dates and return them, no need to mutate the parameters.

File: src/Schedule/WeekdayRule.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    public function excludedDays(\DateTime $from, \DateInterval $period, array $schedule): array
+    {
+        return $schedule;
```
Comment (Magomogo) [inline src/Schedule/WeekdayRule.php]:
```suggestion
        return [];
```

File: src/Schedule/WeekdayRule.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GroupLife\Core\Schedule;
+
+class WeekdayRule implements RuleInterface
```
Comment (Magomogo) [inline src/Schedule/WeekdayRule.php:7]: Please create the test sute for `WeekdayRule`

File: tests/ScheduleTest.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        $oneMonth = new \DateInterval('P1M');
+        $scheduleDays = new Schedule([
+            new Schedule\DayRule(['2021-01-01' => '10:00', '2021-01-02' => '09:00'])
+        ]);
+        $scheduleDays->materialize($dateFrom, $oneMonth);
+
+        $dateCancelFrom = new \DateTime('2021-01-01');
+        $scheduleDays->setRules([
+            new Schedule\CancelDayRule(['2021-01-01' => '10:00'])
+        ]);
```
Comment (Magomogo) [inline tests/ScheduleTest.php:53]: Setter `setRules` is a bad idea, it adds extra complexity without a reason. When you need a new schedule, just call its constructor:

```
$scheduleDays = new Schedule([
    new Schedule\CancelDayRule(['2021-01-01' => '10:00'])
]);

```

**Review by Magomogo — COMMENTED** [review]
> Remove state from Schedule class

Very good commit!

Let's fix the tests here, until build is successful it cannot be merged.

**Pull request merged**